Content.Server/Connection/ConnectionManager.cs

```diff
@@ -175,26 +207,191 @@ private async Task NetMgrOnConnecting(NetConnectingArgs e)
+        }
+
+        public async Task<(bool isWhitelisted, string? denyMessage)> IsWhitelisted(PlayerConnectionWhitelistPrototype whitelist, NetUserData data, ISawmill sawmill)
+        {
+            foreach (var condition in whitelist.Conditions)
+            {
+                var matched = false;
+                string denyMessage;
+                switch (condition.GetType())
+                {
+                    case { } t when t == typeof(ConditionAlwaysMatch):
+                        matched = true;
+                        denyMessage = Loc.GetString("whitelist-always-deny");
+                        break;
+                    case { } t when t == typeof(ConditionManualWhitelist):
+                        matched = !(await _db.GetWhitelistStatusAsync(data.UserId));
+                        denyMessage = Loc.GetString("whitelist-manual");
+                        break;
+                    case { } t when t == typeof(ConditionManualBlacklist):
+                        var blacklisted = await _db.GetBlacklistStatusAsync(data.UserId);
+                        matched = blacklisted;
+                        denyMessage = Loc.GetString("whitelist-blacklisted");
+                        break;
+                    case { } t when t == typeof(ConditionNotesDateRange):
+                        var conditionNotes = (ConditionNotesDateRange)condition;
+                        var remarks = await GetAdminRemarks(data.UserId);
+                        remarks = remarks.Where(x => x.CreatedAt > DateTime.Now.AddDays(-conditionNotes.Range)).ToList();
+                        if (!conditionNotes.IncludeExpired)
+                            // If we're not including expired notes, filter them out.
+                            remarks = remarks.Where(x => x.ExpirationTime is null || x.ExpirationTime > DateTime.Now).ToList();
+                        var remarksCopy = remarks.ToList();
+                        foreach (var adminRemarksRecord in remarks)
+                        {
+                            // In order to get the severity of the remark, we need to see if its a AdminNoteRecord.
+                            if (adminRemarksRecord is not AdminNoteRecord adminNoteRecord)
+                                continue;
+
+                            // We want to filter out secret notes if we're not including them.
+                            if (!conditionNotes.IncludeSecret && adminNoteRecord.Secret)
+                            {
+                                remarksCopy.Remove(adminRemarksRecord);
+                                continue;
+                            }
+
+                            // At this point, we need to remove the note if it's not within the severity range.
+                            if (adminNoteRecord.Severity < conditionNotes.MinimumSeverity)
+                            {
+                                remarksCopy.Remove(adminRemarksRecord);
+                            }
+                        }
```

Comment:
Half this logic is LINQ, the other half isn't. Why can't you just put it all in a single loop with various `continue` blocks for filtering?

You also do not need to bother with the `remarksCopy.Remove()` stuff, if you organize the code well then you can just have a match break the loop with `matched = true;` at the end of it.